recipes_source/generic_join.rst

```diff
@@ -0,0 +1,434 @@
+    ``ZeroRedundancyOptimizer`` together.
+
+Using ``Join`` with ``DistributedDataParallel`` and ``ZeroRedundancyOptimizer``
+-------------------------------------------------------------------------------
+The ``Join`` context manager works not only with a single class but also with
+multiple classes together. PyTorch's ``ZeroRedundancyOptimizer`` is also
+compatible with the context manager, so here, we examine how to modify the
+previous example to use both ``DistributedDataParallel`` and
+``ZeroRedundancyOptimizer``:
+
+::
+
+    from torch.distributed.optim import ZeroRedundancyOptimizer as ZeRO
+    from torch.optim import Adam
+
+    def worker(rank):
+        os.environ['MASTER_ADDR'] = 'localhost'
+        os.environ['MASTER_PORT'] = '29500'
+        dist.init_process_group(BACKEND, rank=rank, world_size=WORLD_SIZE)
+
+        model = DDP(torch.nn.Linear(1, 1).to(rank), device_ids=[rank])
+        optim = ZeRO(model.parameters(), Adam, lr=0.01)
+        # Rank 1 gets one more input than rank 0
+        inputs = [torch.tensor([1]).float() for _ in range(NUM_INPUTS + rank)]
+
+        num_inputs = 0
+        # Pass both `model` and `optim` into `Join()`
+        with Join([model, optim]):
+            for input in inputs:
+                num_inputs += 1
+                loss = model(input).sum()
+                loss.backward()
+                optim.step()
+
+        print(f"Rank {rank} has exhausted all {num_inputs} of its inputs!")
+
+This will yield the same output as before. The notable change was
+additionally passing in the ``ZeroRedundancyOptimizer`` instance into
+``Join()``.
+
+Passing Keyword Arguments
+-------------------------
+Classes may provide keyword arguments that modify their behavior in the context
+manager at run time. For example, ``DistributedDataParallel`` provides an
+argument ``divide_by_initial_world_size``, which determines if gradients are
+divided by the initial world size or by the effective world size (i.e. number
+of non-joined ranks). Such keyword arguments can be passed directly into the
+context manager.
+
+::
+
+    with Join([model, optim], divide_by_initial_world_size=False):
+        for input in inputs:
+            ...
+
+.. warning::
+    The keyword arguments passed into the context manager are shared across
+    all participating classes. This should not be a limitation since we do
+    not expect cases where multiple ``Joinable`` s need differing settings
+    of the same argument. Nonetheless, this is something to keep in mind.
+
+How Does ``Join`` Work?
+-----------------------
+Now that we have seen some preliminary examples of how to use the ``Join``
+context manager, let us delve deeper into how it works. This will provide a
+greater insight into the full capability that it offers and prepare you to make
+your own custom classes compatible. Here, we will go over the ``Join`` class as
+well as the supporting classes ``Joinable`` and ``JoinHook``.
```

Comment:
Let's also make Joinable and JoinHook API docs public first.

Reply:
Done: https://github.com/pytorch/pytorch/pull/62605